은행창구 관리앱 [STEP 1] Blue, Lust3r

llimental opened the pull request:
## BankManager [STEP 1] Blue, Lust3r
안녕하세요! Corn(@protocorn93)🍿!
은행창구 관리 앱 프로젝트를 함께 진행하게 된 💫**TEAM BLUST3R**💫의 Blue, Lust3r입니다!

이번 프로젝트에서 저희가 적용해보고자 하는 바는 아래와 같습니다.
1. **STEP 진행도보다 프로젝트 이해, 학습을 우선시 하여 진행하기**
2. **객체지향 체조 원칙 중 아래 두 항목을 우선 준수하기**
    - 모든 원시값과 문자열을 포장합니다.
    - 3개 이상의 스위프트 기본 데이터 타입(Int, String, Double 등) 프로퍼티를 가진 타입을 구현하지 않습니다.

코드를 작성하기 전에 학습을 우선시 하고 서로 어떻게 진행해야 할지 이야기를 많이 나누어 이해를 바탕으로 프로젝트를 진행하려고 노력했습니다!

아직 부족한 점이 많지만 잘 부탁드립니다! 감사합니다! 🙇🏻🙇🏻

## 🧑🏼‍💻 작업 목록

**[ STEP 1 ] 큐 타입 구현**

- [x] 대기열(Queue) 타입 직접 구현
    - [x] Linked-list 타입 구현
    - [x] Queue 타입에 Generics 기능 활용
    - [x] Unit Test를 통해 Queue 타입 동작 검증
    - [x] 필수 기능(Enqueue, Dequeue, Clear, Peek, isEmpty) 구현

## 📚 학습 키워드 및 학습 내용

### 1. Linked-list
- 노드가 데이터와 포인터를 가지고 한 줄로 연결되어있는 방식으로 데이터를 저장하는 자료구조
- 포인터가 다음이나 이전 노드와의 연결을 담당
- 종류: 단일 연결리스트, 이중 연결리스트, 원형 연결리스트
- 장점: 자료의 추가와 삭제가 $O(1)$의 시간복잡도를 가짐
- 단점: 배열, 트리구조와 달리 특정 위치의 데이터를 검색하는 데에는 $O(n)$의 시간이 걸림

### 2. Queue
- 먼저 집어넣은 데이터가 먼저 나오는 FIFO구조
- 종류: 선형 큐, 환형 큐, 링크드 큐
- 프로젝트상 요구기능 분석
- Enqueue: queue에 값을 추가
- Dequeue: head를 queue에서 제거하고 값을 반환
- Clear: queue에 있는 모든 값을 제거
- Peek: dequeue와 다르게 queue에서 제거하지 않고 head의 값만 반환
- isEmpty: queue가 비어있는지 확인

### 3. Generics
- 타입에 의존하지 않는 범용 코드를 작성할 때 사용
- 서로 다른 Type을 Parameter로 받아 진행할 때, 코드의 재사용성을 높임
- 제네릭 함수
    - 함수 이름 뒤에 `<제네릭타입이름>`을 선언해 주면 해당 이름을 타입처럼 사용할 수 있음
    - `<T>`의 경우 T를 Type Parameter라고 부르는데, 이는 새로운 형식이 생성되는 것이 아니라 실제 함수가 호출될 때 해당 매개변수의 타입으로 대체되는 Placeholder 역할
    - 타입 파라미터의 경우 가독성을 위해 T나 V같은 단일문자, 혹은 Upper Camel Case를 사용

### 4. Unit Test
- 테스트 코드를 작성하는 이유
    - 코드의 안정성을 높일 수 있음
    - 유지보수에 유리함
    - 스펙을 문서화할 수도 있음
    - 깔끔한 코드를 작성할 수 있음

## 🤔 고민과 해결
### 1. Console App의 Unit Test Target 추가
: Unit Test를 위해 프로젝트에서 New Unit Test Target으로 Test Unit을 추가하고 코드를 작성했습니다. 그런데 `@testable import BankManagerConsoleApp` 을 했음에도 undefined symbol 오류가 발생하였습니다.
이유를 찾아본 결과 Application(iOS) 프로젝트의 경우에는 target으로 프로젝트를 선택하여 진행할 수 있지만, command line 프로젝트는 선택할 수 없어 테스트하고자 하는 파일에서 Test Unit에 target membership을 설정해 주어야 했습니다.
테스트를 진행하고자 하는 코드가 있는 LinkedList.swift, LinkedQueue.swift 파일에서 Test Unit을 target membership에서 설정해준 후 문제를 해결할 수 있었습니다.
  

### 2. Queue를 Linked-list로 구현 시의 시간복잡도

: STEP 1이 공개된 후 학습해야 할 키워드를 추출하고, 개인적으로 학습하는 것을 우선으로 진행하였습니다. 그리고 코드 이해도를 위하여 각자 `LinkedQueue`(Linked-list로 구현된 Queue)를 구현해본 후 서로의 코드에 대해 질문하고 논의해 보았습니다.    
논의한 내용을 바탕으로 추가로 이해가 필요한 부분에 대해 함께 학습하고, 기존에 작성한 코드를 전부 삭제한 후 짝 프로그래밍으로 코드를 작성하였습니다.
- 👨🏻‍💻 **Lust3r's Code (단방향 Linked-list)**
    ```swift
    func enqueue(data: T) {
        var head: Node<T>?
        var node = head

        if head == nil {
            head = Node(data: data)

            return
        }

        while node.next != nil {
            node = node.next
        }
        node.next = Node(data: data)
    }
    ```
- 💙 **Blue's Code (단방향 Linked-list + rear 프로퍼티)**
    ```swift
    // LinkedQueue의 enqueue
    mutating func enqueue(data: T) {
        nodes.append(data)
    }

    // LinkedList의 append 기능
    mutating func append(_ data: T) {
        let newNode = Node(data: data)

        count += 1

        if head == nil {
            head = newNode
            tail = head

            return
        }

        tail?.next = newNode
        tail = newNode
    }
    ```
단방향 Linked-list만 사용한 첫 번째 코드는 `rear`(제일 나중에 추가된 노드를 확인하는)프로퍼티가 없어서 `while` 반복문을 사용하여 노드의 `next`가 `nil`이 될 때까지 탐색하는 코드라서 시간복잡도가 $O(n)$이 될 것이라 판단하였습니다.
시간복잡도가 $O(n)$이 된다면, 자료의 추가와 삭제가 $O(1)$의 시간복잡도를 가진다는 Linked-list의 장점이 사라진다고 생각하여 논의 후 새로운 코드를 작성 시 시간복잡도가 $O(1)$이 될 수 있도록 작성해 보았습니다.


## 🙋🏻 질문 거리

### 1. Generic Type 사용 시 Type Parameter 네이밍 질문
- Generic 타입 사용 시 Type Parameter의 경우 가독성을 위해 T나 V같은 단일문자, 혹은 Upper Camel Case를 사용하는 것으로 학습하였고, 저희는 T를 사용하였습니다.    
하지만 위와 같은 네이밍을 피하고 명확히 작성해야 한다는 의견도 있어 Type Parameter를 사용할 때 네이밍을 어떻게 하면 좋을지 궁금합니다!

### 2. Unit Test 진행 시 중복되는 기능에 대한 질문
- 현재 저희가 작성한 코드에서는 `LinkedQueue`의 메서드들이 `LinkedList`의 메서드와 프로퍼티로 구성되어 있습니다. 더 포괄적인 객체인 `LinkedQueue`를 테스트하면 `LinkedList` 또한 테스트가 진행된다고 생각하였습니다.    
혹시 저희 코드처럼 한 객체가 다른 객체의 기능을 활용하는 경우 테스트 케이스가 중복될 위험이 있더라도 각각 Unit Test를 진행해 주는 것이 좋을까요?
(**LinkedList**: `append()`, `removeFirst()`)
(**LinkedQueue**: `enqueue(data:)`, `dequeue()`)

llimental committed:
[docs] DS_Store 파일을 추적하지 않도록 .gitignore 파일 수정

llimental committed:
[feat] Linked-list 타입 구현 및 list에 사용할 Node class 정의

calledBlu committed:
[feat] LinkedList 파일 생성 및 LinkedList를 이용한 LinkedQueue 구현

- 두 파일을 Model로 폴더링

calledBlu committed:
[refactor] LinkedList 구조체 파일 분리 및 이동

- BankManager -> LinkedList로 파일 분리
- LinkedList 파일을 Model로 폴더링

llimental committed:
[refactor] 프로퍼티 네이밍 변경 및 LinkedList 관련 정보를 담는 프로퍼티 선언

- queue -> nodes
- front, rear, count 추가

calledBlu committed:
[refactor] LinkedQueue의 clear()메서드 수정

- mutating 메서드로 변경하면서 불필요해진 self 키워드 제거
- Queue를 Clear한 후 count를 0으로 초기화

calledBlu committed:
[test] LinkedQueue Unit Test 생성 및 진행

- enqueue, dequeue, clear, peek, isEmpty 메서드 테스트 케이스 추가 및 테스트 진행

calledBlu committed:
[refactor] LinkedList의 removeFirst()메서드 로직 개선

- test coverage를 높이기 위한 개선

calledBlu committed:
[test] test coverage 확인 후 테스트되지 않은 코드 수정

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+
+import Foundation
+
+final class Node<T> {
```
`final` 키워드를 붙인 이유에 대해서 설명해주실 수 있으신가요~?

llimental replied:
Node는 다른 클래스에서 상속받아 사용을 하지 않기 때문에 final 키워드를 붙여주게 되었습니다. 이를 통해 Node가 상속이나 override 없이 안정적으로 동작할 수 있다고 생각했습니다.

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+
+final class Node<T> {
+
+    var data: T
```
`data`는 외부에서 읽고 쓸 수 있는데 이 두 행위에 대해서 열어둔 것은 의도하신걸까요~? 의도하신거라면 그 의도가 궁금해요!

calledBlu replied:
LinkedQueue의 peek에서 Node의 data에 접근하고 있기 때문에 private를 쓸 수 없었는데 LinkedQueue의 다른 기능들과 마찬가지로 LinkedList에서 peek의 기능을 구현하고 이 기능을 호출해서 사용한다면 filePrivate로 접근수준을 높일 수 있다고 판단하여 수정하여 반영해 보았습니다!

또한, 저희의 질문에 답변해주셨던 것처럼 data 대신 value로 변수명을 반영해 보았습니다!

> 해당 커밋: 455c53a00a50471f15d5ae0d9a45434fd9d892b7

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+}
+
+struct LinkedList<T> {
```
`LinkedList`를 구조체로 구현한 이유가 궁금해요!

llimental replied:
LinkedList에 대한 이해가 부족했던 것 같습니다. 새로운 노드를 추가하거나 삭제하는 등의 작업을 할 때, 기존에 있던 노드를 참조해야 하기 때문에(next로 다음 노드 참조) class를 사용하는 것이 성능상 좋은데 struct를 사용함으로써 메서드를 호출할 때 값이 복사되는 문제를 일으킬 수 있을 것 같습니다.
더 자료를 찾아본 결과, LinkedList는 struct와 class를 둘 다 사용해도 되지만 데이터를 삽입, 삭제할 때 많은 연산을 필요로 하고, LinkedList의 노드는 다른 노드를 참조해야 하기 때문에 class로 구현하는 것이 더 권장된다고 배우게 되었습니다. 감사합니다!

> 해당 커밋: 895cfaa65e35f282807f2604da9a200c4f2768b8

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+
+struct LinkedList<T> {
+
+    var head: Node<T>?
+    var tail: Node<T>?
```
```swift
var linkedListA = LinkedList() 
linkedListA.head = Node(data: 5)

var linkedListB = linkedListA 
linkedListB.head.data = 100

print(linkedListA.head.data)
print(linkedListB.head.data)
```

이 코드의 결과물은 어떻게 될까요?

calledBlu replied:
저희가 구현한 LinkedList는 struct로 구현되어 있어서 위의 코드를 실행하게 되면 LinkedListA와 LinkedListB가 서로 다른 instance이기 때문에 B에서 head의 데이터를 변경하더라도 A에서 데이터가 변경되지 않는 문제가 발생합니다.
앞선 리뷰에서 LinkedList를 class로 구현했을 때 이점을 바탕으로 LinkedList를 class로 수정하여 해당 코드를 실행한다면 같은 인스턴스를 참조하기 때문에 같은 head의 data를 출력할 수 있게 됩니다.

콘이 직접 작성해주신 예시 코드로 LinkedList를 class로 구현해야 하는 이유를 더 명확하게 알 수 있게 되었어요! 감사해요 콘!

protocorn93 replied:
아뇨아뇨 위와 같이 코드를 작성하게 되면 의도치 않은 변경이 일어날 수 있다는 것을 알려드리고 싶었어요. 

값타입에 기대하는바 중 하나는 불변성이에요. A를 B에 할당했고 B를 변경해도 A는 변경되지 않아야함을 기대할 때 사용하는게 값타입이에요.

하지만 위처럼 값 타입인 구조체가 참조 타입인 클래스 인스턴스 참조 변수를 내부적으로 갖고 있다면, 참조 변수가 복사되기 때문에 B 변경만을 의도했지만 A까지 변경이 되는 현상이 일어날 수 있어요. 

이러한 부분을 유의하면 좋을 것 같다는 의미로 남긴 코멘트였어요!

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+
+    var head: Node<T>?
+    var tail: Node<T>?
+    var count = 0
```
현재 스탭에서 `count`는 필요없어보이는데, 구현하신 이유가 궁금해요!

llimental replied:
LinkedList를 구현하면서 이야기를 나눌 때, 리스트에 있는 노드의 수를 계산해야 하는 상황을 가정했습니다.
count가 없다면 각 노드를 돌면서(node.next) 확인을 해야 하기 때문에 복잡도가 O(n)이라 봤고, 그 때문에 효율적으로 확인하기 위해서는 count를 사용하여 노드 작업 시 체크해주면 좋을 것 같아 구현하였습니다.

하지만 아래에 남겨주신 코멘트와 같이, STEP 1, 그리고 STEP 2에서도 필요한 기능이 아니기 때문에 논의 후 제거했습니다. 감사합니다!

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedQueue.swift:
```diff
@@ -0,0 +1,44 @@
+struct LinkedQueue<T> {
+
+    private var nodes = LinkedList<T>()
+    var front: Node<T>? { nodes.head }
+    var rear: Node<T>? { nodes.tail }
+    var count: Int { nodes.count }
```
이 프로퍼티들은 필요없어보이는데 추가하신 이유가 궁금해요~ 테스트를 위한 코드라면 주어진 인터페이스로만 검증하는게 좋을 것 같아요. `enqueue`후 `dequeue` 결과물로 검증한다던지 말이죠

calledBlu replied:
LinkedList에서 구현된 head, tail, count를 Queue에 구현하기 위해 작성된 프로퍼티였습니다. 기본적으로 구현해야 할 것 같은 큐의 요소들이라 구현해 두었는데 현재 STEP에서는 불필요한 것 같다는 생각이 드네요! 프로퍼티 삭제 후 반영하였습니다!

> 해당 커밋: 02b1ff1bc1db953e8aac6caeb1bbf00d953d1e57

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedQueue.swift:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    func isEmpty() -> Bool {
+        guard nodes.count == 0 else { return false }
```
`count` 프로퍼티 없이 `head`가 `nil` 인지 아닌지로 충분히 검증 가능할 것 같아요

llimental replied:
count 프로퍼티를 삭제한 후 콘이 리뷰해주신 부분을 참고하여 head가 nil인지 아닌지로 검증하는 로직으로 변경하였습니다!

> 해당 커밋: 02b1ff1bc1db953e8aac6caeb1bbf00d953d1e57

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleAppTest/LinkedQueueTest.swift:
```diff
@@ -0,0 +1,122 @@
+        sut = nil
+    }
+
+    func test_enqueue로_데이터가_들어가는지_확인() {
+        sut.enqueue(data: "lust3r")
+
+        let expectation = 1
+
+        XCTAssertEqual(expectation, sut.count)
+    }
```
위에서도 언급했지만 개인적으론 프로퍼티를 통한 검증이 아닌 주어진 인터페이스를 통한 검증이 좋을 것 같아요. `peek` 이라던지, `dequeue`라던지 말이죠.

또한 `count`가 1이라고 해서 `queue`에 들어간 노드의 `data`가 `"lust3r"` 임을 보장하지 않는 것 같아요.

calledBlu replied:
LinkedList와 Queue를 class로 수정하고 불필요한 프로퍼티를 삭제하면서 테스트 케이스의 코드를 전면적으로 수정하였습니다.
콘이 리뷰해주신 것처럼 테스트 케이스를 보장할 수 있도록 테스들을 수정해 보았습니다.

> 해당 커밋: cc5e0e74b4aa3228400e396452269d713e3257ef

protocorn93 commented on BankManagerConsoleApp/BankManagerConsoleApp/Model/LinkedList.swift:
```diff
@@ -0,0 +1,59 @@
+struct LinkedList<T> {
+
+    var head: Node<T>?
+    var tail: Node<T>?
```
`tail`을 활용한 점 좋은 것 같아요! 👍

protocorn93 reviewed (changes requested):
두 분 모두 수고하셨어요! 질문과 코멘트 남겨 놓았어요! 👍

--- 

🙋🏻 질문 거리

> 1. Generic Type 사용 시 Type Parameter 네이밍 질문
> 
> Generic 타입 사용 시 Type Parameter의 경우 가독성을 위해 T나 V같은 단일문자, 혹은 Upper Camel Case를 사용하는 것으로 학습하였고, 저희는 T를 사용하였습니다.
하지만 위와 같은 네이밍을 피하고 명확히 작성해야 한다는 의견도 있어 Type Parameter를 사용할 때 네이밍을 어떻게 하면 좋을지 궁금합니다!

음~ 저는 뭐 이런 네이밍도 크게 나쁘다고 생각이 들진 않아요. 물론 어느정도 역할을 유추할 수 있는 네이밍이 들어가면 좋죠! 이런 경우에는 `Value` 와 같은 네이밍을 사용할 것 같아요. 변수도 `data`보단 `value`가 더 어울리는 것 같구요.

> 2. Unit Test 진행 시 중복되는 기능에 대한 질문
> 
> 현재 저희가 작성한 코드에서는 LinkedQueue의 메서드들이 LinkedList의 메서드와 프로퍼티로 구성되어 있습니다. 더 포괄적인 객체인 LinkedQueue를 테스트하면 LinkedList 또한 테스트가 진행된다고 생각하였습니다.
혹시 저희 코드처럼 한 객체가 다른 객체의 기능을 활용하는 경우 테스트 케이스가 중복될 위험이 있더라도 각각 Unit Test를 진행해 주는 것이 좋을까요?
(LinkedList: append(), removeFirst())
(LinkedQueue: enqueue(data:), dequeue())

LinkedList와 LinkedQueue는 사실 엄연히 다른 객체라고 봐요. LinkedQueue가 LinkedList를 활용할 뿐이죠. LinkedQueue가 내부적으로 다른 형태의 LinkedList를 사용할 수도 있어요. 

이렇게 고민을 하시는 이유는 추상화가 되어 있지 않아서 그렇다고 생각해요. LinkedList와 Queue를 추상화해보고 Queue가 LinkedList를 의존성으로 갖는 형태가 된다면 이 두 객체를 보시는 관점이 달라지실 수 있을 것 같아요. 이에 대해서는 천천히 학습해보셔도 괜찮을 것 같아요!

llimental committed:
[refactor] 프로퍼티 네이밍 변경 및 메서드 수정

- Node의 data -> value
- value의 접근제어 수준을 fileprivate로 설정
- 제네릭 타입 T -> Value
- queue의 첫 값을 보여주는 peek 메서드를 count 사용하지 않고 head를 이용하여 구현

calledBlu committed:
[refactor] 불필요한 프로퍼티 제거 및 관련 메서드 로직 변경, 개행 제거

- LinkedList의 count 프로퍼티 제거
- LinkedQueue의 front, rear, count 프로퍼티 제거
- 위의 프로퍼티 제거에 따른 메서드 내부 변경
- 메서드 중 return 값이 명확하거나 내부의 코드가 한 줄인 경우 축약

llimental committed:
[refactor] LinkedList 타입 및 LinkedQueue 로직 변경

- LinkedList 타입 struct -> class 변경
- 기존 struct LinkedList 에서의 메소드 mutating 키워드 제거
- LinkedQueue에서 LinkedList를 상속함에 따른 메서드 로직 변경

calledBlu committed:
[test] 코드 리팩토링에 따른 테스트코드 수정
